umpost/um2netcdf.py

```diff
@@ -414,6 +365,65 @@ def cubewrite(cube, sman, compression, use64bit, verbose):
         sman.write(cube, zlib=True, complevel=compression, fill_value=fill_value)
 
 
+def fix_forecast_reference_time(cube):
+    # If reference date is before 1600 use proleptic gregorian
+    # calendar and change units from hours to days
+    try:
+        reftime = cube.coord(FORECAST_REFERENCE_TIME)
+        time = cube.coord(TIME)
+        refdate = reftime.units.num2date(reftime.points[0])
+        assert time.units.origin == 'hours since 1970-01-01 00:00:00'
+
+        if time.units.calendar == cf_units.CALENDAR_PROLEPTIC_GREGORIAN and refdate.year < 1600:
+            convert_proleptic(time)
+        else:
+            if time.units.calendar == cf_units.CALENDAR_GREGORIAN:
+                new_calendar = cf_units.CALENDAR_PROLEPTIC_GREGORIAN
+            else:
+                new_calendar = time.units.calendar
+
+            time.units = cf_units.Unit("days since 1970-01-01 00:00", calendar=new_calendar)
+            time.points = time.points / 24.
+
+            if time.bounds is not None:
+                time.bounds = time.bounds / 24.
+
+        cube.remove_coord(FORECAST_PERIOD)
+        cube.remove_coord(FORECAST_REFERENCE_TIME)
+    except iris.exceptions.CoordinateNotFoundError:
+        # Dump files don't have forecast_reference_time
+        pass
+
+
+# TODO: rename time to avoid clash with builtin time module
+def convert_proleptic(time):
+    # Convert units from hours to days and shift origin from 1970 to 0001
+    newunits = cf_units.Unit("days since 0001-01-01 00:00",
+                             calendar=cf_units.CALENDAR_PROLEPTIC_GREGORIAN)
+    tvals = np.array(time.points)  # Need a copy because can't assign to time.points[i]
+    tbnds = np.array(time.bounds) if time.bounds is not None else None
+
+    for i in range(len(time.points)):
+        date = time.units.num2date(tvals[i])
+        newdate = cftime.DatetimeProlepticGregorian(date.year, date.month, date.day,
+                                                    date.hour, date.minute, date.second)
+        tvals[i] = newunits.date2num(newdate)
```

Comment:
I don't think it is necessary to loop through the whole time array. I think it should be ok to just do the `num2date > newdate > date2num` operation on just the first element, work out the difference between this and the original value and apply that offset to the rest of the `tvals` and `tbnds` in two (vectorised) operations.

Reply:
The looping code is old style, conflating logic with manual index fiddling etc. I did some work on #37 to use iteration, but didn't fully solve the problem:
https://github.com/ACCESS-NRI/um2nc-standalone/blob/e50a33a6b3f19050643d02fa5d305ee13d160d8b/umpost/um2netcdf.py#L67-L82